### ISSUE TITLE
Cannot reproduce results by setting random seed

### ISSUE THREAD

**Issue by lguduy** [webdataset/webdataset#122]
Hi. I set random seed for reproducing result, but it seems not work for me.

My code:
```
random.Random(0)
shards = wds.PytorchShardList(file_paths, shuffle=use_shuffle)
dataset = wds.WebDataset(shards).shuffle(size=shuffle_size)
```

I found it set seed in class PytorchShardList:

![image](https://user-images.githubusercontent.com/18639024/140847929-23bfdb6d-159b-4459-af15-39689aed94ae.png)

And set seed in shuffle:

```
shuffle_rng = random.Random()
shuffle_rng.seed((os.getpid(), time.time()))

def shuffle(data, bufsize=1000, initial=100, rng=shuffle_rng, handler=None):
```

How to control random shuffle?

**Comment by tmbdev** [comment]
By "reproducible result" you mean that the sequence of samples you train on differs from run to run?

Here is the current situation:

1. There are two things that need shuffling: shards and samples.
2. In `PytorchShardList`, the shuffle is always reproducible/deterministic but depends on the epoch.
3. The sample shuffle is deterministic if you give it an argument like `rng=random.Random(seed)` 
4. If you use multiple workers, the order in which the `DataLoader` receives samples/batches may be non-deterministic due to concurrency.

So, WebDataset happily hands you reproducible sequences, but DataLoader will then give you non-deterministic batches due to the way its concurrency works for IterableDatasets. You can make this work if you set num_workers=1. You should probably also use SimpleShardList instead of PytorchShardList in that case.

The APIs for all of this have been cleaned up in v2 already.

A general solution will require changing DataLoader or implementing an alternative. That may happen in v2 or v3 or as an add-on.

TODO: reproducible training samples with multiple workers/nodes

**Comment by lguduy** [comment]
I Got it. Thanks for your reply!
